ConfigurationSubstitutor/ConfigurationSubstitutor.cs

```diff
@@ -72,22 +72,21 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
                 if (substitutedValue == null && !string.IsNullOrEmpty(_fallbackDefaultValueDelimiter))
                 {
                     var delimitedVals = capture.Value.Split(new[] { _fallbackDefaultValueDelimiter }, 2, StringSplitOptions.None);
-                    // in case delimiter doesn't match
-                    if (delimitedVals.Length < 2)
+                    // in case delimiter matches
+                    if (delimitedVals.Length == 2)
                     {
-                        return capture.Value;
+                        value = value.Replace(capture.Value, delimitedVals[0]);
+                        // if declared EV isn't resolvable, assign it to provided fallback default value
+                        if (String.IsNullOrEmpty(configuration[delimitedVals[0]]))
+                        {
+                            configuration[delimitedVals[0]] = delimitedVals[1];
```

Comment:
I know this was already added in former PR, but I somehow missed that.
Not a big fan of altering source configuration, if we can keep the function pure and pass the value downwards sounds simpler/better to reason about. What do you think?